### PR TITLE
[RNMobile] Refactor Caption component

### PR DESCRIPTION
**Pull request by geriux**
`Gutenberg-mobile` PR -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/1678

## Description
Our current `Caption` component implementation was meant to be within a block level component, but now there are cases where we just want to use the Caption UI like in the Gallery block.

So this PR adds a new component `BlockCaption` that will have the current logic wrapping a `Caption` component which will only have the UI.

## How has this been tested?
- Open the app
- Add an Image block
  - Set a caption
  - **Expect**: caption being set correctly
- Add a Video block
  - Set a caption
  - **Expect**: caption being set correctly
- Add a Gallery block
  - Add some images
    - Set captions
    - **Expect**: captions being set correctly
  - Set a caption to the Gallery
  - **Expect**: caption being set correctly 
- Save post
- Open it
  - **Expect**: captions are showing correctly

## Types of changes
Refactor

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.
- [x] I've included developer documentation if appropriate.
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/components/caption/index.native.js [hunk 1]
```diff
@@ -1,65 +1,49 @@
...
-		} = select( 'core/block-editor' );
-		const { caption } = getBlockAttributes( clientId );
-		const accessibilityLabel = accessibilityLabelCreator ? accessibilityLabelCreator( caption ) : undefined;
-		const isBlockSelected = getSelectedBlockClientId() === clientId;
+// this platform difference is needed to avoid a regression described here:
+// https://github.com/WordPress/gutenberg/pull/18818#issuecomment-559818548
+const CAPTION_TAG_NAME = Platform.select( {
+	ios: 'p',
+	android: '',
+} );
```
Comment (geriux) [inline packages/block-editor/src/components/caption/index.native.js]: This was in the Gallery block to make it work correctly on Android due to the mentioned regression so I moved it to the Caption component itself since it was affecting the rest of Blocks using it.

Comment (mchowning) [reply packages/block-editor/src/components/caption/index.native.js]: 👍 Linking this up with the issue that was filed: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1651

**Comment by geriux** [comment]
Completed some of the suggested test cases and all good ✅, skipped some not related to these changes.

File: packages/block-editor/src/components/block-caption/index.native.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import { View } from 'react-native';
+
+/**
+ * WordPress dependencies
+ */
+import { Caption, RichText } from '@wordpress/block-editor';
+import { compose } from '@wordpress/compose';
+import { withDispatch, withSelect } from '@wordpress/data';
+
+const BlockCaption = ( { accessible, accessibilityLabel, onBlur, onChange, onFocus, isSelected, shouldDisplay, text } ) => (
```
Comment (mkevins) [inline packages/block-editor/src/components/block-caption/index.native.js]: Nitpick: Could we break this prop list onto multiple lines? It's getting pretty long for one line.

File: packages/block-editor/src/components/block-caption/index.native.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+);
+
+export default compose( [
+	withSelect( ( select, { accessibilityLabelCreator, clientId } ) => {
+		const {
+			getBlockAttributes,
+			getSelectedBlockClientId,
+		} = select( 'core/block-editor' );
+		const { caption } = getBlockAttributes( clientId );
+		const accessibilityLabel = accessibilityLabelCreator ? accessibilityLabelCreator( caption ) : undefined;
```
Comment (mkevins) [inline packages/block-editor/src/components/block-caption/index.native.js]: Not at all a blocker, just something I've been thinking about: I wonder if it makes sense to move this to the caption :thinking: . Also, I wonder if it needs to be set in `withSelect`.. perhaps it can be a regular prop. I understand we get `caption` from block attributes there, but I guess we could also use the `text` that we pass to caption. Wdyt?

File: packages/block-editor/src/components/caption/index.native.js (hunk 1 above)
Comment (mkevins) [reply packages/block-editor/src/components/caption/index.native.js]: I like that this platform divergence is consolidated into one place, OTOH, it seems we currently have to "choose" between two regressions for all captions: [Enter key not working](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1651) or [First word not style-able](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1674#issuecomment-565636271). This PR does not introduce these issues, but we do lose some granularity in how we can "choose which regression" :see_no_evil: for the different use cases. I'm not suggesting a change here, just noting, and hopefully those issues will be resolved, and this will soon be a moot point. :smile:

**Review by mkevins — COMMENTED** [review]
Nice work @geriux ! I've left some minor comments. Overall, this looks pretty good. I've tested this out on Pixel 3a - Android 10, and everything is working well, aside from the known issue with styling the first word of the caption (not introduced by this PR).

File: packages/block-editor/src/components/block-caption/index.native.js (hunk 2 above)
Comment (geriux) [reply packages/block-editor/src/components/block-caption/index.native.js]: I agree! I'll update it =)

File: packages/block-editor/src/components/block-caption/index.native.js (hunk 3 above)
Comment (geriux) [reply packages/block-editor/src/components/block-caption/index.native.js]: Great suggestion! I think we should do it, I'll move some things to the `Caption` component and test how it works with the accessibility changes 😃

File: packages/block-editor/src/components/caption/index.native.js (hunk 1 above)
Comment (geriux) [reply packages/block-editor/src/components/caption/index.native.js]: Yeah you're right, by moving this to the Caption component it will affect Block level components with the [First word not style-able](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1674#issuecomment-565636271) regression, but it will allow users to add new lines on Android... I wonder what users use the most, styling in captions or entering new lines 😅

**Comment by geriux** [comment]
Hi @mkevins 👋 I've updated the PR with the changes from your feedback 😃. I also added the accessibility label for the Video block's caption that was missing.

By the way, I saw that for the images within the gallery we are setting the accessibility label like this -> https://github.com/WordPress/gutenberg/blob/319307c52b35cb168abcb488e9613c5ad4e5e52b/packages/block-library/src/gallery/gallery-image.native.js#L266

Would it make sense to move it to the `Caption` component? Now that it supports to pass the data through its props? What do you think?

Thanks!

**Comment by mkevins** [comment]
Hey @geriux :wave: 

I've tested again with the changes, and everything still works as before. Nice work!

> By the way, I saw that for the images within the gallery we are setting the accessibility label like this ->
> 
> https://github.com/WordPress/gutenberg/blob/319307c52b35cb168abcb488e9613c5ad4e5e52b/packages/block-library/src/gallery/gallery-image.native.js#L266
> 
> Would it make sense to move it to the `Caption` component? Now that it supports to pass the data through its props? What do you think?

I tried that on Android, and it seemed to be ok, but I haven't had a chance to check on iOS. Since the comment specifically mentions addressing a VoiceOver issue for iOS, perhaps it's best to leave it for now. The intent here is to treat an unselected `GalleryImage` as a "leaf", i.e. no children elements are talked out on the VoiceOver, so moving this to Caption may regress that.

**Review by mkevins — APPROVED** [review]
I had to locally merge with latest (gutenberg:master and gutenberg-mobile:develop) to the respective branches for these PRs due to an [intermittent dependency issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1703), but after that, everything is working nicely. Nice work @geriux :+1: . After updating to latest, LGTM!

**Comment by geriux** [comment]
> I had to locally merge with latest (gutenberg:master and gutenberg-mobile:develop) to the respective branches for these PRs due to an [intermittent dependency issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1703), but after that, everything is working nicely. Nice work @geriux 👍 . After updating to latest, LGTM!

Awesome! Thanks for testing it!

**Comment by geriux** [comment]
Hey @pinarol ! This is ready to be merged but I want to double check this https://github.com/WordPress/gutenberg/pull/19118/files/865739561806ba1a735f161d40df17f935ddf00c#diff-fb4bb656dce8b784120b171a9b46cbcaR12-R17 first.

**TL;DR** With these changes we'll have one regression for captions **(only on Android)** which is that the first word can't be styled because we are not setting the `CAPTION_TAG_NAME`. The reason we are not setting it is because of another regression that doesn't allow adding [new lines](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1651).

Currently, for **(Android)** we have two regressions in captions:

- `Gallery block caption` User can add new lines but can't style first word of the caption
- `Other captions` User can't add new lines but can style all words of the caption.

What do you think? Should we merge and have just one regression for all captions? Or wait until those bugs are fixed?

**Comment by pinarol** [comment]
@geriux So we need to choose between bugs :) I gave it a try and I think not being able to style feels more buggy than not being able to enter a new line. When I select the whole sentence and try to apply a style it is simply not responding. So I'd go with sacrificing the ability to add a new line.

**Comment by geriux** [comment]
> @geriux So we need to choose between bugs :) I gave it a try and I think not being able to style feels more buggy than not being able to enter a new line. When I select the whole sentence and try to apply a style it is simply not responding. So I'd go with sacrificing the ability to add a new line. Plus, I don't expect multiple line captions are that common.

Sounds good! I agree with multiple line captions not being that common, awesome! I'll update it then =)

Thank you!

**Comment by geriux** [comment]
Hey @mkevins I changed this https://github.com/WordPress/gutenberg/pull/19118/commits/1359e55d1d6b4b662d1d1f31cfba923477d90659 if it looks good to you I can go ahead and merge =)

**Comment by mkevins** [comment]
Hi @geriux :wave: This looks good to merge. Nice work!
